### DISCUSSION TITLE
Moving away from doctests

### DISCUSSION THREAD

**Discussion by lthoerner** [ratatui/ratatui#324] [Discussion]
Unit tests are great, and I am fully on-board with TDD in general. However, due to the lack of LSP/editor support for doctests, and the library's pervasive reliance on said tests, I find myself frequently having to manually refactor large swaths of tests in a way that, if the tests were not in their LSP-avoidant format, would take only a few keystrokes.

I understand having examples, but I do not feel that tests *in* the examples has enough utility to justify this ongoing inconvenience. In fact, I think that the requirement of importing everything manually so the code actually compiles is detrimental to the examples, insofar as it adds in a lot of extraneous information that could easily be inferred by the user (very few users are going to say "oh, I need to import this struct before I use it?").

I suggest we refactor existing doctests into unit test submodules, and going forward only include code above methods, structs, and other items as examples, rather than unit tests. This should be supplemented by ongoing development of the Ratatui book. Alternatively, we can move all doctests to a dedicated section of the `tests/` directory, though I don't see much utility in this.

**Comment by joshka** [comment]
I'm in agreement with the sentiment (we should prefer generally to write unit tests rather than doc tests).

It's important to keep in mind that good documentation shows how a feature is used well enough that a user of the library shouldn't have to guess or read the code to work out what the normal case looks like. Asserts in a doc test often are used as a shortcut for explaining exactly what the result of a certain call looks like where a textual description would be longer and have more ambiguity.

One thing that is good about doc tests is that when there is a lot of complexity in the setup, you might be finding that there is an actual problem with the complexity of the API.

Doing a quick grep over `/// .*assert` only brought two obvious places that I would subjectively choose to rewrite as unit tests (`Block::inner()` and `enum Color`). Can you point out any specific pain points that you hit that I'm missing? It's possible that some other changes have helped reduce some of this code (like introducing a `::new()` method in `StyledGrapheme`, and perhaps using `prelude::*` / `widgets::*` would help reduce some of the noise.

Do you think it's worth doing this across the entire repo, or would you just make it a general rule to do this cleanup work as part of any feature work?

**Reply by lthoerner** [reply]
Almost all the doctests have one or more `assert!()`s, the grep should be over `/// .*assert.*`. The biggest problem with doctests is their resistence to refactoring tools. For instance, renaming a type will not rename it in doctests. Unless there is a VSCode extension I am unaware of, there is no way to do this outside of find/replace or manual searching.

I think this should be done in one large PR, though this is debatable, I'm open to having these conversions performed upon editing their respective areas in the codebase.

**Reply by joshka** [reply]
yup - I looked at all the doctests. Can you give a concrete example perhaps?

Renaming - find and replace via regex is a pretty useful tool in your toolbox to have.
e.g.:
Find: `///(.*)SomeType`
Replace: `///$1SomeOtherType`

**Reply by lthoerner** [reply]
Concrete example - having to rename `Spans` to `Line`.

Regex replacement with capture groups is nice, and I use it occasionally, but it shouldn't be relied on too much.

**Reply by joshka** [reply]
That's in the past though. I want to understand the problem that you are solving today with this. I'm finding it a bit difficult to generalize to the problem and hence the solution, without a singular example to extrapolate from.

**Reply by lthoerner** [reply]
The problem is in the future when we want to do refactoring without needing a lot of manual work. Another problem I forgot to mention is that you will not see compiler errors in doctests until you run them, which wastes unnecessary time where you have to hunt down errors in commented code rather than just seeing inline errors in your tests module.

Also, keep this in mind.
> *I think that the requirement of importing everything manually so the code actually compiles is detrimental to the examples, insofar as it adds in a lot of extraneous information that could easily be inferred by the user ...*

**Reply by joshka** [reply]
Aside from the two examples I gave (Block and Color), what is the best example you're aware of that shows this problem?
Can you point out the first change that you'd make?

**Reply by lthoerner** [reply]
Both the issues listed in my previous comment are pervasive - that is to say, they occur in close to every doctest. The former is an issue with the formatting of doctests, not with their readability. Write a compiler error in a doctest and you'll notice that it won't be caught by the language server until you try to run it.

If you really must have a specific example of the latter, here is one.

```rs
    /// Patches the style of each Span in an existing Spans, adding modifiers from the given style.
    ///
    /// ## Examples
    ///
    /// ```rust
    /// # use ratatui::text::{Span, Spans};
    /// # use ratatui::style::{Color, Style, Modifier};
    /// let style = Style::default().fg(Color::Yellow).add_modifier(Modifier::ITALIC);
    /// let mut raw_spans = Spans::from(vec![
    ///     Span::raw("My"),
    ///     Span::raw(" text"),
    /// ]);
    /// let mut styled_spans = Spans::from(vec![
    ///     Span::styled("My", style),
    ///     Span::styled(" text", style),
    /// ]);
    ///
    /// assert_ne!(raw_spans, styled_spans);
    ///
    /// raw_spans.patch_style(style);
    /// assert_eq!(raw_spans, styled_spans);
```
would be much more readable and concise as
```rs
/// let style = Style::default().fg(Color::Yellow).add_modifier(Modifier:ITALIC);
/// let mut unstyled_spans = Spans::from(vec![
///     Span::raw("My"),
///     Span::raw(" text"),
/// ]);
///
/// raw_spans.patch_style(style);
```

Sorry to sound petulant, but I'm not really sure where the disconnect is happening here. Almost all the issues I've mentioned are not specific to any one part of the code, or task being performed on the codebase. I fail to see how extensive examples are helpful to a dialogue as general as this one.

**Reply by TimerErTim** [reply]
Hey, guys, it's me again

Just my 50 cents: 
I see doctests as a simple way to ensure that the provided code examples in the docs are actually working as expected (compiling, returning the expected results, etc.) so that we don't show library users wrong (legacy) usage.

With that in mind, I agree with @lthoerner , because the examples should ve very concise and only includes bare minimum and low level asserts to ensure future changes don't break the according doc example.

So moving complete test cases out of the docs and into a deticated module (or somewhere else) seems to intuitively make the most sense.

It won't fix the refactoring problem due to bad LSP/VS-Code support regarding doctests. Which in turn makes them even more important to ensure we didn't miss some example during a refactor. 

One would have to look through the whole  code base and extract all overly verbose and sophisticated tests in doc examples into their new according module.
Or change this gradually by clearly documenting these new doctest guidelines so that future contributions don't overengineer the doctests.

With all that being said, I think the majority of docs perfectly utilize asserts and don't require any rework at all.

**Comment by joshka** [comment]
I'm in agreement with both of you that doctests shouldn't have a lot of complexity, nor should they be the place where we rely on to define all the behavior of the method.

I also agree that the `Spans` example (and the equivalent on `Line`) is more complex than it needs to be. However, its important to understand why the `Spans` doc test is useful. Most configuration / builder style methods in Ratatui just set a value and return self. This is one place where we take a different approach. It's important to show the effect of this method, and not just how to call it. With that in mind, I'd simplify the example to:

```rust
    /// ```rust
    /// use ratatui::{text::{Span, Spans}, style::{Style, Stylize}};
    /// let mut spans = Spans::from("foo");
    /// spans.patch_style(Style::new().red());
    /// assert_eq!(spans.0[0].style, Style::new().red());
    /// ```
```

This communicates several things succinctly:
- `patch_style` mutates the `Spans` instance.
- `patch_style` applies the change directly to the interior `Span`s. It is not deferred to being applied while rendering. This is important to note as any `Span` added afterward does not inherit the patched style. To communicate that better for readers of the docs, we should probably add:

```rust
    /// // The style is patched, not replaced.
    /// spans.0.push(Span::raw("bar"));
    /// assert_ne!(spans.0[1].style, Style::new().red());
```

(Incidentally, I think this feature is probably better handled by storing a style field and then applying that at render time, which avoids this complexity altogether)

English prose is one way to describe this behavior, but it carries inherent ambiguity in doing so. Code does not have the same ambiguity. It either runs or it doesn't.

You're also right about LSP support for comments. It's annoying that this is lacking, but doc comments are not for the library developer, they are for the library user. I generally find that having a pane running `bacon test` while I'm coding can help highlight issues with tests early. Using find and replace as a complementary tool to refactoring is often a necessary part of development regardless of how powerful your refactoring tools are.

Like @TimerErTim says, I'm not sure that this is generally a problem that requires removing, but it's would be helpful to think about how we write new and maintain existing documentation. I think it's important to make this a clear part of our contribution guidelines that, of the three test types, unit tests are the most important, and how we'd prefer doc tests to be written. Here's the current testing section. Perhaps you have some suggestions on how to simplify / reword this to cover those points?

https://github.com/tui-rs-revival/ratatui/blob/daf589015290ac8b379389d29ef90a1af15e3f75/CONTRIBUTING.md?plain=1#L92-L114

Regarding the disconnect, it's easy to overgeneralize a problem in software and I think that's happened a bit in this discussion. Some problems with over-generalization:
- It does not apply as widely as stated.
- It does not consider other alternative explanations or approaches (similar to [The XY Problem](https://xyproblem.info/).
- It might be valid, but it's not something that has a high enough impact to be worth spending time on.

Starting with a few examples and then extrapolating helps mitigate these problems, and makes it easier to understand the generalization.

# TL;DR:
Concretely I'd suggest:
- Keep writing doc tests. They are generally useful when done right.
- Fix any complex doc tests that are causing specific pain
- Propose some better wording for the CONTRIBUTING document
- Propose some tasks to add to the PR template
- Propose some ways to improve the docs generally
  - I like using LLMs for generation of docs. They need clear guidelines to provide useful output.

**Comment by joshka** [comment]
I think https://github.com/ratatui-org/ratatui/pull/371 is a good example of when asserts in doctests are useful for showing edge cases.
